refactor(core): data block handling, normalize message formats, strip IDs from messages

mdrxy opened the pull request:
> [!WARNING]
> **BREAKING:** Simplifies message normalization to single consistent path, requiring partner package updates

**Key Changes:**
- Consistent multimodal handling:
  - OpenAI `image_url` blocks pass through unchanged (broad compatibility)
  - OpenAI `input_audio` and `file` blocks convert to v1 standard equivalents
  - Legacy v0 multimodal blocks convert to v1 standard
  - Everything else passes through unchanged
- Partner packages must update content block parsing logic

**Partner Updates**

`output_version` affects how messages are serialized into `.content`. `_normalize_messages()` will now upgrade v0 content to v1, so, all partners now receive v1 format input regardless of `output_version`.

Migration:
- Partner packages must update to handle v1 input content blocks
- `output_version` still controls serialization format of responses (unchanged)

mdrxy committed:
refactor: enhance OpenAI data block handling and normalize message formats

mdrxy committed:
Merge branch 'cc/1.0/standard_content' into mdrxy/normalize

mdrxy committed:
Merge branch 'cc/1.0/standard_content' into mdrxy/normalize

mdrxy committed:
import

mdrxy committed:
fix(messages): handle OpenAI-specific file content blocks in `content_blocks`

mdrxy committed:
Merge branch 'cc/1.0/standard_content' into mdrxy/normalize

mdrxy committed:
fix(tests): update content block tests for handling missing 'type' keys

mdrxy committed:
ss

mdrxy committed:
ss

mdrxy committed:
Merge branch 'cc/1.0/standard_content' into mdrxy/normalize

mdrxy committed:
Merge branch 'cc/1.0/standard_content' into mdrxy/normalize

mdrxy committed:
Merge branch 'cc/1.0/standard_content' into mdrxy/normalize

mdrxy committed:
enhance

mdrxy committed:
.

mdrxy committed:
Merge branch 'cc/1.0/standard_content' of github.com:langchain-ai/langchain into mdrxy/normalize

mdrxy committed:
.

mdrxy committed:
.

mdrxy committed:
Add message normalization for caching to remove auto-generated IDs

mdrxy committed:
.

mdrxy committed:
undo formatting

mdrxy committed:
.

mdrxy committed:
.

mdrxy committed:
lint

mdrxy committed:
v0 normalize

mdrxy committed:
Add parameterization to rate limit cache tests for output version

mdrxy committed:
Update content block conversion to handle audio and file types

ccurme commented on libs/core/langchain_core/messages/base.py:
```diff
@@ -129,19 +134,54 @@ def content_blocks(self) -> list[types.ContentBlock]:
+                    "image_url",
+                    "input_audio",
+                    "file",
+                } and _is_openai_data_block(item):
```
can we refactor this logic into helper functions:
```python
def _convert_to_v1_from_chat_completions_input(content: list[types.ContentBlock]) -> list[types.ContentBlock]:

def _convert_to_v1_from_v0_multimodal_input(content: list[types.ContentBlock]) -> list[types.ContentBlock]:
```
so that we can do something like:
```python
@property
def content_blocks(self) -> list[types.ContentBlock]:
    blocks: list[types.ContentBlock] = []

    # First pass converting to standard blocks
    content = (
        [self.content]
        if isinstance(self.content, str) and self.content
        else self.content
    )
    for item in content:
        if isinstance(item, str):
            blocks.append({"type": "text", "text": item})
        elif isinstance(item, dict):
            item_type = item.get("type")
            if item_type not in types.KNOWN_BLOCK_TYPES:
                blocks.append({"type": "non_standard", "value": item})
            else:
                blocks.append(cast("types.ContentBlock", item))
        else:
            pass

    # Subsequent passes attempt to unpack non-standard blocks
    blocks = _convert_to_v1_from_v0_multimodal_input(blocks)
    blocks = _convert_to_v1_from_chat_completions_input(blocks)
    blocks = _convert_to_v1_from_anthropic_input(blocks)
    blocks = _convert_to_v1_from_gemini_input(blocks)
    ...

    return blocks
```
open to other ways of organizing this.

ccurme reviewed (commented):
Let's assume that we will update all chat models that accept v0 mulltimodal blocks to take v1.

I'm not sure there's a way around this— if a user sets `output_version = "v1"`, then we will normalize to v1 types anyway, right? the purpose of `output_version` IMO is to control what is serialized, nothing should break if it's flipped.

So I think it's OK to not have a separate `_normalize_messages_v0` path. `_normalize_messages` should
- pass through image_url blocks
- update chat completions file and audio blocks to v1
- update v0 image, audio, file blocks to v1
- pass through everything else unchanged

mdrxy commented on libs/core/langchain_core/language_models/_utils.py:
```diff
@@ -57,43 +89,204 @@ def _parse_data_uri(uri: str) -> Optional[dict]:
     if block["type"] == "input_audio":
-        data = block["input_audio"].get("data")
-        audio_format = block["input_audio"].get("format")
+        audio = block.get("audio", {})
```
good catch

mdrxy committed:
resolve some comments

mdrxy committed:
fix test

mdrxy committed:
resolving more comments

mdrxy committed:
ai -> human

mdrxy committed:
rfc to use _convert_legacy_v0_content_block_to_v1

mdrxy committed:
rfc reducing the number of type ignores

mdrxy committed:
Add utility function to compare content blocks while ignoring `id` fields

mdrxy committed:
more rfc

mdrxy committed:
rfc

mdrxy committed:
.

mdrxy committed:
.

mdrxy committed:
docstrings

mdrxy committed:
remove alternate v0 path

mdrxy committed:
feat: `_strip_ids_from_messages` (#32622)

Fixes #31949

**Root Cause:**
- Dynamically generated unique IDs in message arrays were preventing
cache hits (cache key was based on the full serialized message
structure)
- These IDs changed on each run, making the serialized input different
every time
- Even when the semantic content was identical, the cache key would be
different due to these varying IDs

**Impact:**

- Cache misses on identical semantic content - The same logical
conversation would trigger new API calls
- More API calls than necessary

**Fix:**
- Remove all IDs depending on `strip_all_ids` in
`_strip_ids_from_messages` - False: Only removes IDs starting with
`LC_AUTO_PREFIX`, True: Removes ALL IDs regardless of prefix for maximum
cache hit potential

ccurme reviewed (commented):
Thanks for this, will look more tomorrow morning.

mdrxy commented on libs/core/langchain_core/language_models/_utils.py:
```diff
@@ -57,90 +97,354 @@ def _parse_data_uri(uri: str) -> Optional[dict]:
+
+                    # Convert OpenAI audio/file block to LC v1 std content
+                    # unless `all` is True, in which case we also conver images
+                    if convert_all or block["type"] != "image_url":
```
Not yet, but gives a path forward in the future if needed

mdrxy committed:
remove redundant append

mdrxy committed:
move func

mdrxy committed:
fix: improve v0 content block conversion handling

mdrxy committed:
fix: typing

ccurme commented on libs/core/tests/unit_tests/language_models/chat_models/test_base.py:
```diff
@@ -461,19 +493,86 @@ def test_trace_images_in_openai_format() -> None:
             "url": "https://example.com/image.png",
         }
     ]
+    # But .content_blocks should perform v1 transformation
```
this test can go in tests/unit_tests/messages/block_translators/test_openai.py

ccurme commented on libs/core/tests/unit_tests/language_models/chat_models/test_base.py:
```diff
@@ -461,19 +493,86 @@ def test_trace_images_in_openai_format() -> None:
+            ]
+        ]
+    ]
+    # With v1 output_version, .content should be transformed
+    # Check structure, ignoring auto-generated IDs
+    assert len(response.content) == 1
+    content_block = response.content[0]
+    if isinstance(content_block, dict) and "id" in content_block:
+        # Remove auto-generated id for comparison
+        content_without_id = {k: v for k, v in content_block.items() if k != "id"}
+        expected_content = {
+            "type": "image",
+            "url": "https://example.com/image.png",
+        }
+        assert content_without_id == expected_content
+    else:
+        assert content_block == {
+            "type": "image",
+            "url": "https://example.com/image.png",
+        }
```
my sense is that the changes to ParrotFakeChatModel above stem from hijacking this test to test content block translation in addition to tracing.

ccurme commented on libs/core/tests/unit_tests/language_models/chat_models/test_base.py:
```diff
@@ -508,156 +607,247 @@ def test_trace_content_blocks_with_no_type_key() -> None:
+        ]
+    )
+
+    normalized_content = _normalize_messages([messages], convert_all=True)
```
I would not add `convert_all` until we use it somewhere. right now it costs complexity for no benefit.

https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it

ccurme commented on libs/core/tests/unit_tests/language_models/chat_models/test_base.py:
```diff
@@ -679,18 +869,24 @@ def test_output_version_invoke(monkeypatch: Any) -> None:
     assert response.content == [{"type": "text", "text": "hello"}]
     assert response.response_metadata["output_version"] == "v1"
 
-    llm = GenericFakeChatModel(messages=iter(messages))
-    response = await llm.ainvoke("hello")
-    assert response.content == "hello"
```
(nit) ordering was deliberate here

ccurme commented on libs/core/tests/unit_tests/language_models/chat_models/test_base.py:
```diff
@@ -701,33 +897,20 @@ async def test_output_version_ainvoke(monkeypatch: Any) -> None:
         assert isinstance(chunk.content, str)
         assert chunk.content
-        full = chunk if full is None else full + chunk
+        full = chunk if full is None else full + chunk  # type: ignore[assignment]
```
why is this type ignore now necessary?

unclear to me what improvements we're making to the tests here.

ccurme committed:
remove id stripping

mdrxy commented on libs/core/langchain_core/language_models/_utils.py:
```diff
@@ -57,90 +96,348 @@ def _parse_data_uri(uri: str) -> Optional[dict]:
+        image_url_known_keys = {"url"}
+        image_url_extras = _extract_extras(block["image_url"], image_url_known_keys)
+
+        all_extras = {**extras}
```
?

mdrxy committed:
remove unnecessary final condition

mdrxy committed:
undo and switch import

mdrxy committed:
relocate some things

mdrxy committed:
remove `convert_all`

mdrxy committed:
add test for v0 to v1 content block transformation

mdrxy committed:
parrot

mdrxy committed:
special case required for parrot in test_trace_images_in_openai_format_v0

mdrxy committed:
nvm

mdrxy committed:
Merge branch 'cc/1.0/standard_content' into mdrxy/normalize

mdrxy committed:
lint

mdrxy committed:
.
